File: app/widgets/SegmentManager.py
import string
from typing import List
import cv2 as cv
import numpy as np
from PySide6.QtSvgWidgets import QSvgWidget
from PySide6.QtWidgets import *
from PySide6.QtGui import QAction, QPaintEvent, QPixmap, QTransform, QImage, QMouseEvent
from PySide6.QtCore import Slot, Qt, Signal
from PySide6 import QtCore
from widgets.ImageLabel import ImageLabel
from src.opencv_qt_compat import *
from assets import bse_segments_assets


class SegmentManager(QWidget):
    swapped = Signal(int, int)

    def __init__(self):
        super().__init__()
        self.imgLabels: List[ImageLabel] = []
        self.imgs = []
        self.moving = None

        # self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Expanding)

        for i in range(4):
            self.imgLabels.append(ImageLabel())

        layout = QGridLayout()
        layout.addWidget(self.imgLabels[0], 0, 0, 2, 2)
        layout.addWidget(self.imgLabels[1], 0, 4, 2, 2)
        layout.addWidget(self.imgLabels[2], 2, 0, 2, 2)
        layout.addWidget(self.imgLabels[3], 2, 4, 2, 2)

        segmentImage = QSvgWidget(":/segments/bse_detector.svg")
        segmentImage.renderer().setAspectRatioMode(Qt.KeepAspectRatio)
        segmentImage.setMinimumSize(100, 100)
        segmentImage.setMaximumSize(150, 212)
        segmentImage.setSizePolicy(QSizePolicy.Minimum, QSizePolicy.Minimum)
        # segmentImage.setScaledContents(True)
        layout.addWidget(segmentImage, 1, 2, 2, 2)

        layout.setAlignment(segmentImage, Qt.AlignCenter)
        # layout.setColumnStretch(0, 300)
        # layout.setColumnStretch(1, 300)
        # layout.setColumnStretch(4, 300)
        # layout.setColumnStretch(5, 300)
        self.setLayout(layout)

    def loadImages(self, imgs):
        if (len(imgs) != 4):
            return

        self.imgs = imgs
        for i in range(4):
            self.imgLabels[i].setPixmap(mat2PixRGB(self.imgs[i]), 100, 100)

    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        if event.buttons() & Qt.LeftButton:
            if self.moving == None:
                x = event.x()
                y = event.y()
                xwidg = self.width()
                ywidg = self.height()
                self.moving = x//(xwidg//2) + (y//(ywidg//2))*2

    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        if event.button() == Qt.LeftButton:
            if self.moving != None:
                moving = self.moving
                self.moving = None
                x = event.x()
                y = event.y()
                xwidg = self.width()
                ywidg = self.height()
                target = x//(xwidg//2) + (y//(ywidg//2))*2
                if target < 0 or target == moving:
                    return
                self.swapImages(moving, target)

    def swapImages(self, first: int, second: int):
        if (not self.imgs):
            return
        tmp = self.imgs[first]
        self.imgs[first] = self.imgs[second]
        self.imgs[second] = tmp
        for i in range(4):
            self.imgLabels[i].replacePixmap(mat2PixRGB(self.imgs[i]))

        self.swapped.emit(first, second)

File: app_c++/BakalarkaTake2/widgets/SegmentManager/SegmentManager.cpp
#include "SegmentManager.h"
#include <QSvgRenderer>

SegmentManager::SegmentManager(QWidget* parent)
	: QWidget(parent)
{
	setupUi(this);
	segmentImage->load(QString(":/gui/svg/bse_detector.svg"));
	segmentImage->renderer()->setAspectRatioMode(Qt::KeepAspectRatio);

	m_labels = { imageLabel0, imageLabel1, imageLabel2, imageLabel3 };
}

SegmentManager::~SegmentManager()
{
}

void SegmentManager::loadImages(std::array<cv::Mat, 4> imgs)
{
	for (int i = 0; i < 4; i++) {
		m_labels[i]->setMat(imgs[i]);
	}
}

void SegmentManager::mouseMoveEvent(QMouseEvent* event)
{
	if (event->buttons() & Qt::LeftButton) {
		if (m_moving == NOT_MOVING) {
			auto x = event->pos().x();
			auto y = event->pos().y();
			auto xwidg = width();
			auto ywidg = height();
			m_moving = x * 2 / xwidg + y * 2 / ywidg * 2;
		}
	}
}

void SegmentManager::mouseReleaseEvent(QMouseEvent* event)
{
	if (event->button() == Qt::LeftButton) {
		if (m_moving != NOT_MOVING) {
			auto x = event->pos().x();
			auto y = event->pos().y();
			auto xwidg = width();
			auto ywidg = height();
			auto target = x * 2 / xwidg + y * 2 / ywidg * 2;

			if (target < 0 || target == m_moving) {
				return;
			}

			swapImages(m_moving, target);
			m_moving = NOT_MOVING;
		}
	}
}

void SegmentManager::swapImages(int first, int second)
{
	assert(first >= 0 && first < 4);
	assert(second >= 0 && second < 4);
	assert(first != second);

	auto tmp = m_labels[first]->fullPixmap();
	m_labels[first]->setPixmap(m_labels[second]->fullPixmap());
	m_labels[second]->setPixmap(tmp);
	emit swapped(first, second);
}
